evm-test-helpers/src/contracts/oracle.ts

```diff
@@ -0,0 +1,317 @@
+ * @param callbackFunctionId The callback function id for the response
+ * @param nonce The nonce sent by the requester
+ * @param data The CBOR payload of the request
+ */
+export function encodeOracleRequest(
+  specId: string,
+  callbackAddr: string,
+  callbackFunctionId: string,
+  nonce: number,
+  data: BigNumberish,
+): string {
+  const oracleRequestSighash = '0x40429946'
+  const oracleRequestInputs = [
+    { name: '_sender', type: 'address' },
+    { name: '_payment', type: 'uint256' },
+    { name: '_specId', type: 'bytes32' },
+    { name: '_callbackAddress', type: 'address' },
+    { name: '_callbackFunctionId', type: 'bytes4' },
+    { name: '_nonce', type: 'uint256' },
+    { name: '_dataVersion', type: 'uint256' },
+    { name: '_data', type: 'bytes' },
+  ]
+
+  const encodedParams = ethers.utils.defaultAbiCoder.encode(
+    oracleRequestInputs.map(i => i.type),
+    [
+      ethers.constants.AddressZero,
+      0,
+      specId,
+      callbackAddr,
+      callbackFunctionId,
+      nonce,
+      1,
+      data,
+    ],
+  )
+
+  return `${oracleRequestSighash}${stripHexPrefix(encodedParams)}`
+}
+
+/**
+ * Extract a javascript representation of a run request from the data
+ * contained within a EVM log.
+ * ```solidity
+ *  event OracleRequest(
+ *    bytes32 indexed specId,
+ *    address requester,
+ *    bytes32 requestId,
+ *    uint256 payment,
+ *    address callbackAddr,
+ *    bytes4 callbackFunctionId,
+ *    uint256 cancelExpiration,
+ *    uint256 dataVersion,
+ *    bytes data
+ *  );
+ * ```
+ *
+ * @param log The log to extract the run request from
+ */
+export function decodeRunRequest(log?: ethers.providers.Log): RunRequest {
+  if (!log) {
+    throw Error('No logs found to decode')
+  }
+
+  const types = [
+    'address',
+    'bytes32',
+    'uint256',
+    'address',
+    'bytes4',
+    'uint256',
+    'uint256',
+    'bytes',
+  ]
+  const [
+    requester,
+    requestId,
+    payment,
+    callbackAddress,
+    callbackFunc,
+    expiration,
+    version,
+    data,
+  ] = ethers.utils.defaultAbiCoder.decode(types, log.data)
+
+  return {
+    specId: log.topics[1],
+    requester,
+    requestId: toHex(requestId),
+    payment: toHex(payment),
+    callbackAddr: callbackAddress,
+    callbackFunc: toHex(callbackFunc),
+    expiration: toHex(expiration),
+    data: addCBORMapDelimiters(Buffer.from(stripHexPrefix(data), 'hex')),
+    dataVersion: version.toNumber(),
+
+    topic: log.topics[0],
+  }
+}
+
+/**
+ * Extract a javascript representation of a ConcreteChainlinked#Request event
+ * from an EVM log.
+ * ```solidity
+ *  event Request(
+ *    bytes32 id,
+ *    address callbackAddress,
+ *    bytes4 callbackfunctionSelector,
+ *    bytes data
+ *  );
+ * ```
+ * The request event is emitted from the `ConcreteChainlinked.sol` testing contract.
+ *
+ * @param log The log to decode
+ */
+export function decodeCCRequest(
+  log: ethers.providers.Log,
+): [string, string, string, string] {
```

Comment:
How do you feel about extracting semantic type names for these even though they're all strings? e.g.

```
export function decodeCCRequest(
  log: ethers.providers.Log,
): [RequestId, Address, CallbackFunctionSelector, Data] {
```

Reply:
I think that'd get really messy, ideally we would have nominal types but typescript doesn't support them by default. The reason why I embedded the solidity snippet is so that users can see what's returned to them by referencing the jsdoc. We could even return a typed object like
```ts
interface {
  string id,
  string callbackAddress,
  string callbackfunctionSelector,
  string data
}
```
but I think that would be out of scope for now, trying to minimize breaking changes

Reply:
👍 typed object would be ideal but can be done separately